rust/src/sdp/parser.rs

```diff
@@ -0,0 +1,694 @@
+    let (i, first_digit) = peek(be_u8)(i)?;
+    if first_digit == 0 {
+        return Err(Err::Error(make_error(i, ErrorKind::HexDigit)));
+    }
+    let (i, num) = take_u8(i)?;
+    Ok((i, num))
+}
+
+// SDP Message format (fields marked with * are optional):
+// https://www.rfc-editor.org/rfc/rfc4566#page-9
+//
+//       Session description
+//         v=  (protocol version)
+//         o=  (originator and session identifier)
+//         s=  (session name)
+//         i=* (session information)
+//         u=* (URI of description)
+//         e=* (email address)
+//         p=* (phone number)
+//         c=* (connection information -- not required if included in
+//              all media)
+//         b=* (zero or more bandwidth information lines)
+//         One or more time descriptions ("t=" and "r=" lines; see below)
+//         z=* (time zone adjustments)
+//         k=* (encryption key)
+//         a=* (zero or more session attribute lines)
+//         Zero or more media descriptions
+//
+//      Time description
+//         t=  (time the session is active)
+//         r=* (zero or more repeat times)
+//
+//      Media description, if present
+//         m=  (media name and transport address)
+//         i=* (media title)
+//         c=* (connection information -- optional if included at
+//              session level)
+//         b=* (zero or more bandwidth information lines)
+//         k=* (encryption key)
+//         a=* (zero or more media attribute lines)
+
+pub fn sdp_parse_message(i: &[u8]) -> IResult<&[u8], SdpMessage> {
+    let (i, version) = parse_version_line(i)?;
+    let (i, origin) = parse_origin_line(i)?;
+    let (i, session_name) = parse_session_name(i)?;
+    let (i, session_info) = opt(parse_session_info)(i)?;
+    let (i, uri) = opt(parse_uri)(i)?;
+    let (i, email) = opt(parse_email)(i)?;
+    let (i, phone_number) = opt(parse_phone_number)(i)?;
+    let (i, connection_data) = opt(parse_connection_data)(i)?;
+    let (i, bandwidths) = opt(parse_bandwidth)(i)?;
+    let (i, time) = parse_time(i)?;
+    let (i, repeat_time) = opt(parse_repeat_times)(i)?;
+    let (i, time_zone) = opt(parse_time_zone)(i)?;
+    let (i, encryption_key) = opt(parse_encryption_key)(i)?;
+    let (i, attributes) = opt(parse_attributes)(i)?;
+    let (i, media_description) = opt(many0(parse_media_description))(i)?;
+    Ok((
+        i,
+        SdpMessage {
+            version,
+            origin,
+            session_name,
+            session_info,
+            uri,
+            email,
+            phone_number,
+            connection_data,
+            bandwidths,
+            time,
+            repeat_time,
+            time_zone,
+            encryption_key,
+            attributes,
+            media_description,
+        },
+    ))
+}
+
+fn parse_version_line(i: &[u8]) -> IResult<&[u8], u32> {
+    let (i, _) = tag("v=")(i)?;
+    let (i, _v) = tag("0")(i)?;
+    let (i, _) = line_ending(i)?;
+
+    Ok((i, 0))
```

Comment:
Why store it if it is always 0 ?

Reply:
I wanted to use `msg.version` as a reference within the logger, instead of using `0` directly.

Reply:
But why log it if it is always 0 ?

Reply:
I wanted to follow the RFC, but I'm ok with removing it.

Reply:
I would either
- parse an arbitrary integer and log it
- just check `v=0` and not log nor store anything